matrix-api.c

```diff
@@ -933,6 +933,51 @@ MatrixApiRequestData *matrix_api_download_thumb(MatrixConnectionData *conn,
     return fetch_data;
 }
 
+MatrixApiRequestData *matrix_api_upload_keys(MatrixConnectionData *conn,
+        JsonObject *device_keys, JsonObject *one_time_keys,
+        MatrixApiCallback callback,
+        MatrixApiErrorCallback error_callback,
+        MatrixApiBadResponseCallback bad_response_callback,
+        gpointer user_data)
+{
+    GString *url;
+    MatrixApiRequestData *fetch_data;
+    JsonNode *body_node;
+    JsonObject *top_obj;
+    JsonGenerator *generator;
+    gchar *json;
+
+    url = g_string_new(conn->homeserver);
+    g_string_append(url, "_matrix/client/unstable/keys/upload?access_token=");
+    g_string_append(url, purple_url_encode(conn->access_token));
+
+    top_obj = json_object_new();
```

Comment:
Done, I unref this now after I do the json_node_set_object with it (which
inc's it's ref)  Note this changes a few semantics, in that the keys that get passed to this upload get freed.